### ISSUE TITLE
SignedCms compatibility for the SignerInfo SignatureAlgorithm issue

### ISSUE THREAD

**Issue by vcsjones** [dotnet/corefx#32639]
I've observed another divergence of behavior from `SignedCms` between netcoreapp2.1 and net472.

If the CMS has a SignerInfo with a `SignatureAlgorithm` OID of "1.2.840.113549.1.1.11" (sha256WithRSAEncryption), an exception is raised:

>CryptographicException: 'Unknown algorithm '1.2.840.113549.1.1.11'.'

It appears that `CmsSignature` is only expecting '1.2.840.113549.1.1.1' which is just RSA Encryption:

https://github.com/dotnet/corefx/blob/0ce22c97d7117d2cf974e1da40bb4ba72fa53e65/src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/CmsSignature.RSA.cs#L18-L19

This behavior diverges with the Desktop .NET Framework (CAPI) which is tolerant of SignerInfos with a signature algorithm OID 1.2.840.113549.1.1.11 (and presumably 1.2.840.113549.1.1.12, etc for SHA384).

I've put together a small repro issue over at https://github.com/vcsjones/SignedCmsBug.

This succeeds:

```
dotnet run --framework net472
```

this fails:

```
dotnet run --framework netcoreapp2.1
```

This is using .NET Core 2.1.5.

/cc @bartonjs

**Comment by onovotny** [comment]
FWIW, Bouncy Castle appears to use the sha256WithRSAEncryption OID here. That's how I ran into it.

**Comment by onovotny** [comment]
@bartonjs wouldn't this need to hit 2.1.x as an LTS release too?

**Comment by onovotny** [comment]
Also, related, and likely the same root cause, calling `cms.Encode()` after loading it via `Decode(data)`, on the message fails with corrupt ASN.1 message exception.

**Comment by bartonjs** [comment]
@onovotny That's what I originally thought, too.  But apparently LTS means the opposite of that.

> Large organisations, or users with mission critical projects, need the security fixes but would often prefer to retain the same base version for an extended period without any new or changed functionality.

<sub>https://en.wikipedia.org/wiki/Long-term_support#Rationale</sub>

So it'd need a scenario that justifies the risk of change.

**Comment by onovotny** [comment]
@bartonjs how about existing stuff in the wild isn't validating on it? Does that count?

**Comment by onovotny** [comment]
Actually isn't System.Security.Cryptography.Pkcs shipped as an out-of-band NuGet package anyway? If that's serviced, can't users pull in a later version with the fix?

**Comment by bartonjs** [comment]
> Actually isn't System.Security.Cryptography.Pkcs shipped as an out-of-band NuGet package anyway?

Sort of.  It's out of band for Microsoft.NETCore.App.  But it's included in Microsoft.AspNetCore.App.

**Comment by vcsjones** [comment]
@bartonjs I can do a PR for adding those OIDs if you agree that is the right thing to do for 3.0 while we sort out if this is right for 2.1.x. Thoughts?

**Comment by bartonjs** [comment]
@vcsjones It should follow the DSA model, where it makes sure that if (e.g.) sha1WithRsaEncryption is specified that the digest algorithm says SHA-1.

https://github.com/dotnet/corefx/blob/1d737919b3a74dd3d26a80b2268014b48a41a8d4/src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/CmsSignature.DSA.cs#L14-L21

**Comment by vcsjones** [comment]
@onovotny 

>Also, related, and likely the same root cause, calling cms.Encode() after loading it via Decode(data), on the message fails with corrupt ASN.1 message exception.

Unfortunately this fix does not address that issue which we saw from the BouncyCastle signature. I'll dig in to it and file a separate issue if necessary when I'm able to track down what the cause it.

<details>
<pre>
        System.Security.Cryptography.CryptographicException : ASN1 corrupted data.
        Stack Trace:
          C:\dev\Personal\corefx\src\Common\src\System\Security\Cryptography\AsnReader.cs(213,0): at System.Security.Cryptography.Asn1.AsnReader.ReadTagAndLength(Nullable`1& contentsLength, Int32& bytesRead)
          C:\dev\Personal\corefx\src\Common\src\System\Security\Cryptography\AsnReader.cs(295,0): at System.Security.Cryptography.Asn1.AsnReader.PeekEncodedValue()
          C:\dev\Personal\corefx\src\Common\src\System\Security\Cryptography\AsnReader.cs(336,0): at System.Security.Cryptography.Asn1.AsnReader.GetEncodedValue()
          C:\dev\Personal\corefx\src\Common\src\System\Security\Cryptography\AsnWriter.cs(183,0): at System.Security.Cryptography.Asn1.AsnWriter.WriteEncodedValue(ReadOnlyMemory`1 preEncodedValue)
          C:\dev\Personal\corefx\src\System.Security.Cryptography.Pkcs\src\System\Security\Cryptography\Pkcs\Asn1\EncapsulatedContentInfoAsn.xml.cs(28,0): at System.Security.Cryptography.Pkcs.Asn1.EncapsulatedContentInfoAsn.Encode(AsnWriter writer, Asn1Tag tag)
          C:\dev\Personal\corefx\src\System.Security.Cryptography.Pkcs\src\System\Security\Cryptography\Pkcs\Asn1\EncapsulatedContentInfoAsn.xml.cs(16,0): at System.Security.Cryptography.Pkcs.Asn1.EncapsulatedContentInfoAsn.Encode(AsnWriter writer)
          C:\dev\Personal\corefx\src\System.Security.Cryptography.Pkcs\src\System\Security\Cryptography\Pkcs\Asn1\SignedDataAsn.xml.cs(37,0): at System.Security.Cryptography.Pkcs.Asn1.SignedDataAsn.Encode(AsnWriter writer, Asn1Tag tag)
          C:\dev\Personal\corefx\src\System.Security.Cryptography.Pkcs\src\System\Security\Cryptography\Pkcs\Asn1\SignedDataAsn.xml.cs(21,0): at System.Security.Cryptography.Pkcs.Asn1.SignedDataAsn.Encode(AsnWriter writer)
          C:\dev\Personal\corefx\src\System.Security.Cryptography.Pkcs\src\System\Security\Cryptography\Pkcs\SignedCms.cs(120,0): at System.Security.Cryptography.Pkcs.SignedCms.Encode()
</pre>
</details>

**Comment by karelz** [comment]
Ported into release/2.1 in PR #33463. Will be part of 2.1.7 release.

**Comment by InternetLogic** [comment]
I'm a little confused as to whether the initial issue is fixed or won't fix? 
We are experiencing the following behaviour which departs from .net4 behaviour in both netcore2.x and netcore3:
Calling SignedCMS.CheckSignature on a perfectly valid signature throws an exception for the following algorithms:
SHA256RSA (OID "1.2.840.113549.1.1.11")
SHA384RSA (OID: "1.2.840.113549.1.1.12")
SHA512RSA (OID: "1.2.840.113549.1.1.13")
SHA1RSA (OID: "1.2.840.113549.1.1.5")
The CryptographicException thrown is either:
"'_<OID>_'is not a known hash algorithm."
or 
"Unknown algorithm '_<OID>_'."

**Comment by vcsjones** [comment]
@internetlogic 

Should be fixed. Can you confirm you are using the latest stable version of the S.S.C.Pkcs NuGet package (4.5.2)? 

Can you provide a signed CMS that reproduces the behavior?

**Comment by InternetLogic** [comment]
DohI New to .NET Core and I'm using CLR! I'll raise it there

**Comment by InternetLogic** [comment]
> @InternetLogic
> 
> Should be fixed. Can you confirm you are using the latest stable version of the S.S.C.Pkcs NuGet package (4.5.2)?
> 
> Can you provide a signed CMS that reproduces the behavior?

Wasn't it fixed as of 2.1.7? In which case shouldn't it be present in the current installable framework?

**Comment by vcsjones** [comment]
The `S.S.C.Pkcs` package is a bit weird. As far as I understand, it is not an in-the-box component for `Sdk="Microsoft.NET.Sdk"` projects. You need to add a nuget package. If you are using ASP.NET Core, it's an in-the-box component.  Earlier up in the discussion:

>>Actually isn't System.Security.Cryptography.Pkcs shipped as an out-of-band NuGet package anyway?

>Sort of. It's out of band for Microsoft.NETCore.App. But it's included in Microsoft.AspNetCore.App.

https://github.com/dotnet/corefx/issues/32639#issuecomment-427430981

It would still be helpful to make sure you are on the right version of everything, and to get a CMS that can reproduce the issue.

**Comment by InternetLogic** [comment]
@vcsjones I have investigated this further I've created a test that shows the divergent behaviour. I'm using the latest NuGet S.S.C.Pkcs netcore2.2 and netcore3. The signature is created and validates on net4-net4.7.2. In netcore 2.2 and netcore3 it throws "1.2.840.113549.1.1.11 is not a known hash algorithm." CryptographicException. Here's a project - there's a .p12 in the bin/debug. I've also included a signeddata.cms. Though if you run the same code in net4.x it will generate it for you. 
[TestSignedCms.zip](https://github.com/dotnet/corefx/files/3321071/TestSignedCms.zip)

**Comment by vcsjones** [comment]
@InternetLogic thanks for the report. This is a slightly different case that appears to be a duplicate of #28504.

@bartonjs the problem is at the snippet below, and would easy to add additional OIDs. I agree with the assessment of #28504 that these OIDs are technically "wrong" in so much that they are not really digest algorithm identifiers. However, if a CMS in the wild has this identifier, it is not verifiable on Core, though is on Desktop.

If this is something you think should be fixed, I'll open a new issue for this change and implement it.

https://github.com/dotnet/corefx/blob/2861ef06530df06b70f17a91616d979c8f18f75e/src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/PkcsHelpers.cs#L39-L52

**Comment by InternetLogic** [comment]
Incidentally the documents we are signing are PDFs and Acrobat recognises these Oids and validates the signatures. Should it not be possible to use S.S.S.Pkcs we'll probably have to use bouncy castle for validation.

**Comment by bartonjs** [comment]
@vcsjones I think I can accept reading the intent out of existing files, but I think I'd prefer that the signature algorithm OIDs not be considered valid when creating new signatures.  (i.e. the code in #28504 still fails, but this one succeeds).

So probably means adding `bool forVerify` (or whatever) to `GetDigestAlgorithm`.

For best chances of being fixed in 3.0, the PR needs to be merged in about 51 hours (or less).

**Comment by vcsjones** [comment]
>the PR needs to be merged in about 51 hours (or less).

On it.
